### PR TITLE
feat: Adds support for SSL to address issue #71.

### LINKED ISSUE #71: Support for SSL
This is technically already supported via the sqlalchemy_url configuration. But we should add config option(s) for SSL as well so you don't have to use `sqlalchemy_url` to support this. 

Also as a part of this we should note in the `sqlalchemy_url` and `dialect+driver` that these settings are not safe to use as they can change significantly in the future, they should be used in combination with creating a ticket going over why you had to use the `sqlalchemy_url` and could not make the original configuration settings work.

### PR DESCRIPTION
**Pull request by sebastianswms**
Two types of SSL are supported:

- Checking the server's certificate against a certificate authority using `ssl_enable`, for verifying the server's identity to the client.
- Passing a client certificate and private key to the server using `ssl_client_certificate_enable`, for verifying the client's identity to the server.

Each can be configured independently of the other. Using either or both precludes the use of `sqlalchemy_url`.

Closes #71.

### CODE REVIEW

File: target_postgres/connector.py [hunk 1]
```diff
@@ -174,3 +176,76 @@ def get_column_add_ddl(
...
+            A dictionary with key-value pairs for the sqlalchemy query
+
+        """
+        if path.isfile(value):
+            return {key: value}
+        else:
+            with open(alternative_name, "wb") as alternative_file:
+                alternative_file.write(value.encode("utf-8"))
+            if restrict_permissions:
+                chmod(alternative_name, 0o600)
```
Comment (visch) [inline target_postgres/connector.py:250]:
```suggestion
                chmod(alternative_name, 0x600)
```

Does this work?  Just looks odd to be 0o

**Review by visch — CHANGES_REQUESTED** [review]

File: .github/workflows/release_workflow.yml [hunk 2]
```diff
@@ -38,4 +38,4 @@ jobs:
     - name: Deploy to PyPI
       run: |
         poetry publish -r testpypi -u "__token__" -p "${{ secrets.TEST_PYPI_TOKEN }}"
-        poetry publish -u "__token__" -p "${{ secrets.POSTGRES_PYPI_TOKEN }}"
\ No newline at end of file
+        poetry publish -u "__token__" -p "${{ secrets.POSTGRES_PYPI_TOKEN }}"
```
[the remark this answers is not in the document]
Comment (sebastianswms) [reply .github/workflows/release_workflow.yml:41]: I failed a linting pre-commit check; yes. https://results.pre-commit.ci/run/github/513737416/1684442449.tCJ2HgSBSdKi71SuThdRWA

File: target_postgres/target.py [hunk 3]
```diff
@@ -138,6 +167,70 @@ def __init__(
...
+                + " authentication to the server. Use ssl_certificate and"
+                + " ssl_private_key for further customization. To use SSL to verify"
+                + " the server's identity, use ssl_enable instead."
+                + " Note if sqlalchemy_url is set this will be ignored."
+            ),
+        ),
+        th.Property(
+            "ssl_mode",
+            th.StringType,
+            default="require",
```
[the remark this answers is not in the document]
Comment (sebastianswms) [reply target_postgres/target.py:197]: I've switched the default to be `verify-full`.

File: target_postgres/connector.py [hunk 4]
```diff
@@ -174,3 +176,76 @@ def get_column_add_ddl(
...
+        # ssl_client_certificate_enable is for verifying the client's identity to the server.
+        if config["ssl_client_certificate_enable"]:
+            query.update(
+                self.filepath_or_certificate(
+                    "sslcert", config["ssl_certificate"], ".secrets/cert.crt"
+                )
+            )
+            query.update(
+                self.filepath_or_certificate(
+                    "sslkey", config["ssl_private_key"], ".secrets/pkey.key", True
```
[the remark this answers is not in the document]
Comment (sebastianswms) [reply target_postgres/connector.py:213]: No, I don't think so. `alternative_name` is not a default in the event the user fails to provide a value, it's the location that a file should be created if the user provides a raw certificate instead of a filepath. If the user provides a raw certificate (regardless of the default value), there still needs to be a file to store the certificate, which is when `alternative_name` is used.

File: target_postgres/target.py [hunk 5]
```diff
@@ -96,8 +125,8 @@ def __init__(
             description=(
                 "SQLAlchemy connection string. "
                 + "This will override using host, user, password, port, "
-                + "dialect. Note that you must esacpe password special "
-                + "characters properly see "
+                + "dialect, and all ssl settings. Note that you must escape password "
```
[the remark this answers is not in the document]
Comment (sebastianswms) [reply target_postgres/target.py:128]: I never encountered a problem with it. That notice was present before I started on this PR.

It looks like we already escape special characters for users if provided for the `password` config option, just not when they use an SQLAlchemy URL. I don't think we should attempt to modify an SQLAlchemy URL since it creates ambiguity.

When a user passes an SQLAlchemyURL of `dialect+driver://username:alpha@bravo:1234/charlie@delta:4321/echo`, should that be resolved as
```
password=alpha
host=bravo
port=1234
database=charlie%40delta%3A4321%2Fecho
```
or
```
password=alpha%40bravo%3A1234%2Fcharlie
host=delta
port=4321
database=echo
```

The [SQLAlchemy documentation](https://docs.sqlalchemy.org/en/20/core/engines.html#escaping-special-characters-such-as-signs-in-passwords) asks for special characters to be escaped, and I think we should trust users to follow that specification.

File: target_postgres/target.py [hunk 6]
```diff
@@ -138,6 +167,70 @@ def __init__(
...
+            ),
+        ),
+        th.Property(
+            "ssl_client_certificate_enable",
+            th.BooleanType,
+            default=False,
+            description=(
+                "Whether or not to provide client-side certificates as a method of"
+                + " authentication to the server. Use ssl_certificate and"
+                + " ssl_private_key for further customization. To use SSL to verify"
```
[the remark this answers is not in the document]
Comment (sebastianswms) [reply target_postgres/target.py:189]: The current way it's set up, there are two settings to enable/disable the two types of SSL—`ssl_enable` and `ssl_client_certificate_enable`—with each having two settings that allow further customization: `ssl_mode` and `ssl_certificate_authority` for the former, `ssl_certificate` and `ssl_private_key` for the latter.

If you think it's desirable, I could change it so that the two enable/disable settings are removed, and the use or nonuse of SSL is determined by whether the other settings have values or not.

File: target_postgres/target.py [hunk 7]
```diff
@@ -47,6 +47,35 @@ def __init__(
             + "password, and dialect+driver to be set"
         )
 
+        # If sqlalchemy_url is not being used and ssl_enable is on, ssl_mode must have one of six allowable values. If ssl_mode is verify-ca or verify-full, a certificate authority must be provided to verify against.
+        assert (
```
[the remark this answers is not in the document]
Comment (sebastianswms) [reply target_postgres/target.py:53]: Tests have been added.

File: .github/workflows/ci_workflow.yml [hunk 8]
```diff
@@ -33,23 +33,19 @@ jobs:
...
-          - 5432:5432
     strategy:
       matrix:
         python-version: ["3.7", "3.8", "3.9", "3.10", "3.11"]
     steps:
     - uses: actions/checkout@v3
+    - name: Fix key permissions
+      run: |
+        chmod 600 ssl/server.key
+        sudo chown 999:999 ssl/server.key
```
Comment (visch) [inline .github/workflows/ci_workflow.yml:44]: uid/gid 999? Do you know why it's that group and user?

File: .github/workflows/release_workflow.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (visch) [reply .github/workflows/release_workflow.yml]: got it, we'll just leave it

File: .github/workflows/ci_workflow.yml (hunk 8 above)
Comment (sebastianswms) [reply .github/workflows/ci_workflow.yml:44]: I was getting permission issues, and when I checked the uid of the postgres user, 999:999 was displayed. Switched it to that and the permission issues cleared up.

File: target_postgres/target.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (visch) [reply target_postgres/target.py:131]: Let's leave it, I thought you added those words, it looks like it was as lint issue so I"ll just pass on thinking about it, I thought you hit something which is why there was a change here

File: target_postgres/tests/test_standard_target.py [hunk 9]
```diff
@@ -397,3 +421,125 @@ def test_activate_version_deletes_data_properly(postgres_config):
...
+    The server is configured with certificates that state it is hosted at "localhost",
+    which won't match the loopback address "127.0.0.1". Because verify-full (the
+    default) requires them to match, an error is expected.
+    """
+    tap = SampleTapCountries(config={}, state=None)
+
+    postgres_config_modified = copy.deepcopy(postgres_config)
+    postgres_config_modified["host"] = "127.0.0.1"
+
+    with pytest.raises(sqlalchemy.exc.OperationalError):
```
Comment (visch) [inline target_postgres/tests/test_standard_target.py:494]: In this case I'm not too worried about it, but an FYI for later is normally when you catch an Exception you want to be specific about what kind of exception it is. In this case `OperationalError` isn't very specific so you'd have to check that the message matches the type of Exception you're expecting. I think this is alright to leave as is though!

File: target_postgres/tests/test_standard_target.py [hunk 10]
```diff
@@ -397,3 +421,125 @@ def test_activate_version_deletes_data_properly(postgres_config):
...
+    # for when ssl_client_certificate_enable is on.
+    with pytest.raises(AssertionError):
+        TargetPostgres(config=postgres_config_modified)
+
+
+def test_postgres_ssl_invalid_cn(postgres_config):
+    """Test that connection will fail due to non-matching common names.
+
+    The server is configured with certificates that state it is hosted at "localhost",
+    which won't match the loopback address "127.0.0.1". Because verify-full (the
```
Comment (visch) [inline target_postgres/tests/test_standard_target.py:486]: /nit 
Instead of a depdendency on the global setting being set to `verify-full` for this test we should just set it in the test. 

ie

``` postgres_config_modified["ssl_mode"] = "verify-full"`

Just so this test is more self contained, and we don't lose anything by doing this here.

File: target_postgres/tests/test_standard_target.py [hunk 11]
```diff
@@ -397,3 +421,125 @@ def test_activate_version_deletes_data_properly(postgres_config):
     with engine.connect() as connection:
         result = connection.execute(f"SELECT * FROM {table_name}")
         assert result.rowcount == 0
+
+
+def test_postgres_ssl_no_config(postgres_config_no_ssl):
```
Comment (visch) [inline target_postgres/tests/test_standard_target.py:426]: Great tests!

**Review by visch — CHANGES_REQUESTED** [review]
Looks really good, just a few things. Also I'm going to merge a seperate PR and we may get some conflicts here that you'll need to address

File: target_postgres/tests/test_standard_target.py (hunk 9 above)
Comment (sebastianswms) [reply target_postgres/tests/test_standard_target.py:494]: Just for my future reference—isn't checking the error message of the Exception a bit fragile? That seems like something that could easily change.

File: target_postgres/tests/test_standard_target.py (hunk 10 above)
Comment (sebastianswms) [reply target_postgres/tests/test_standard_target.py:486]: I made this change.

File: target_postgres/target.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (visch) [reply target_postgres/target.py:193]: I think I get it now but it's really confusing to me.  In the readme can we add a section on SSL and describe all of this in more detail.

Also if `ssl_client_certificate_enable` is only used with `ssl_certificate` and `ssl_private_key` then should we name it `ssl_client_certificate` and `ssl_client_private_key` , It's also surprising to me 

I think https://www.postgresql.org/docs/current/libpq-ssl.html does a good job explaining all of this so maybe a link to here would help too?

File: target_postgres/connector.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (visch) [reply target_postgres/connector.py]: Why would we default here instead of `~/.postgresql/postgresql.key` as listed in https://www.postgresql.org/docs/current/libpq-ssl.html

File: target_postgres/tests/test_standard_target.py (hunk 9 above)
Comment (visch) [reply target_postgres/tests/test_standard_target.py:511]: Right it's better to have a specific exception for what you're trying to catch / verify. If you don't have that (as is the case here I believe) you have to use the message which is fragile, but at least you know you're catching the right exception as `OpertationalError` gets thrown for a ton of reasons

**Review by visch — CHANGES_REQUESTED** [review]
Few doc tweaks / config name tweaks then we're good I think

**Comment by sebastianswms** [comment]
>  In the readme can we add a section on SSL and describe all of this in more detail.

Done.

> Why would we default here instead of `~/.postgresql/postgresql.key`

If that's where the user has their certificate and private key, they can simply pass that file path. But if the user explicitly passes in a raw certificate (as opposed to a path to a file with a certificate), the raw certificate needs to be put into a file before it can be used for authentication. Overwriting the user's postgresql certificate and key seems like dangerous behavior, so a local directory inside the tap is used instead.

In short: I can picture a situation where a user has a value in `~/.postgresql/postgresql.key` that they do not want overwritten *and* that they also don't want to use for the purposes of the tap.

> but at least you know you're catching the right exception

Makes sense. Thanks.

File: target_postgres/connector.py [hunk 12]
```diff
@@ -222,3 +224,82 @@ def get_column_add_ddl(
...
+                    key="sslcert",
+                    value=config["ssl_certificate"],
+                    alternative_name=".secrets/cert.crt",
+                )
+            )
+            query.update(
+                self.filepath_or_certificate(
+                    key="sslkey",
+                    value=config["ssl_private_key"],
+                    alternative_name=".secrets/pkey.key",
```
Comment (visch) [inline target_postgres/connector.py:266]: I think we need to allow folks to change this as sometimes this directory isn't writable by folks. I think it's a good default though we should just add a configuration for this

File: target_postgres/connector.py [hunk 13]
```diff
@@ -222,3 +224,82 @@ def get_column_add_ddl(
...
+                self.filepath_or_certificate(
+                    key="sslcert",
+                    value=config["ssl_certificate"],
+                    alternative_name=".secrets/cert.crt",
+                )
+            )
+            query.update(
+                self.filepath_or_certificate(
+                    key="sslkey",
+                    value=config["ssl_private_key"],
```
Comment (visch) [inline target_postgres/connector.py:265]: From the conversation https://github.com/MeltanoLabs/target-postgres/pull/131#issuecomment-1599552075 , I think my point is the default for `ssl_private_key` (set in `tap.py`) should be `~/.postgresql/postgresql.key` which I think solves the question you're asking. 

If you're confused give me a buzz!

**Review by visch — CHANGES_REQUESTED** [review]

File: target_postgres/target.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (visch) [reply target_postgres/target.py:193]: @sebastianswms I think we still need this as well

File: target_postgres/connector.py [hunk 14]
```diff
@@ -222,3 +224,82 @@ def get_column_add_ddl(
...
+            )
+        return query
+
+    def filepath_or_certificate(
+        self,
+        key: str,
+        value: str,
+        alternative_name: str,
+        restrict_permissions: bool = False,
+    ) -> dict[str, str]:
```
[the remark this answers is not in the document]
Comment (sebastianswms) [reply target_postgres/connector.py:278]: Done.

**Review by visch — APPROVED** [review]

**Review by visch — CHANGES_REQUESTED** [review]
Linting is failing

**Review by visch — APPROVED** [review]
